### PR TITLE
Reduce verbosity of the bulk indexing audit log

### LINKED ISSUE #66568: Cut down audit verbosity for bulk index actions
I think it's probably better to illustrate the issue with two examples that show the audited output of bulk requests:

1) given the following index request:
```
curl -u elastic:password -X POST "localhost:9200/_bulk?pretty" -H 'Content-Type: application/json' -d'
{ "index" : { "_index" : "test", "_id" : "1" } }
{ "field1" : "value1" }
{ "index" : { "_index" : "test2", "_id" : "1" } }
{ "field1" : "value1" }
{ "index" : { "_index" : "test3", "_id" : "1" } }
{ "field1" : "value1" }

{ "delete" : { "_index" : "test", "_id" : "2" } }
{ "delete" : { "_index" : "test2", "_id" : "2" } }
{ "delete" : { "_index" : "test3", "_id" : "2" } }

{ "create" : { "_index" : "test", "_id" : "2" } }
{ "field1" : "value2" }
{ "create" : { "_index" : "test2", "_id" : "2" } }
{ "field1" : "value2" }
{ "create" : { "_index" : "test3", "_id" : "2" } }
{ "field1" : "value2" }

{ "update" : {"_id" : "1", "_index" : "test"} }
{ "doc" : {"field2" : "value2"} }
{ "update" : {"_id" : "1", "_index" : "test2"} }
{ "doc" : {"field2" : "value2"} }
{ "update" : {"_id" : "1", "_index" : "test3"} }
{ "doc" : {"field2" : "value2"} }
'
```
, which performs all the 4 allowed bulk index action types on each of 3 distinct indices (test, test2, and test3),
it will produce the following simplified audit output:
```
{ ..."event.action":"access_granted", ... , "origin.type":"rest", ... , "action":"indices:data/write/bulk", "request.name":"BulkRequest"}

{ ... "event.action":"access_granted", ... , "action":"indices:data/write/bulk[s]", "request.name":"BulkShardRequest", "indices":["test2"]}
{ ... "event.action":"access_granted", ... , "action":"indices:data/write/index:op_type/index", "request.name":"BulkItemRequest", "indices":["test2"]}
{ ..., "event.action":"access_granted", ... , "action":"indices:data/write/delete", "request.name":"BulkItemRequest", "indices":["test2"]}
{ ..., "event.action":"access_granted", ... , "action":"indices:data/write/index:op_type/create", "request.name":"BulkItemRequest", "indices":["test2"]}
{..., "event.action":"access_granted", ... , "action":"indices:data/write/update", "request.name":"BulkItemRequest", "indices":["test2"]}
{..., "event.action":"access_granted", ... , "action":"indices:data/write/bulk[s][p]", "request.name":"BulkShardRequest", "indices":["test2"]}

{..., "event.action":"access_granted", ... , "action":"indices:data/write/bulk[s]", "request.name":"BulkShardRequest", "indices":["test3"]}
{..., "event.action":"access_granted", ... , "action":"indices:data/write/index:op_type/index", "request.name":"BulkItemRequest", "indices":["test3"]}
{..., "event.action":"access_granted", ... , "action":"indices:data/write/delete", "request.name":"BulkItemRequest", "indices":["test3"]}
{..., "event.action":"access_granted", ..., "action":"indices:data/write/index:op_type/create", "request.name":"BulkItemRequest", "indices":["test3"]}
{..., "event.action":"access_granted", ..., "action":"indices:data/write/update", "request.name":"BulkItemRequest", "indices":["test3"]}
{..., "event.action":"access_granted", ..., "action":"indices:data/write/bulk[s][p]", "request.name":"BulkShardRequest", "indices":["test3"]}

{..., "event.action":"access_granted", ... , "action":"indices:data/write/bulk[s]", "request.name":"BulkShardRequest", "indices":["test"]}
{..., "event.action":"access_granted", ..., "action":"indices:data/write/index:op_type/index", "request.name":"BulkItemRequest", "indices":["test"]}
{..., "event.action":"access_granted", ..., "action":"indices:data/write/delete", "request.name":"BulkItemRequest", "indices":["test"]}
{..., "event.action":"access_granted", ... , "action":"indices:data/write/index:op_type/create", "request.name":"BulkItemRequest", "indices":["test"]}
{..., "event.action":"access_granted", ... , "action":"indices:data/write/update", "request.name":"BulkItemRequest", "indices":["test"]}
{..., "event.action":"access_granted", ... , "action":"indices:data/write/bulk[s][p]", "request.name":"BulkShardRequest", "indices":["test"]}
```
The particular verbosity that I take issue with in this case is that each operation type has a separate entry for each index name.

I suggest we collate the entries representing the `BulkItemRequest`, such that a single entry has multiple actions. The exact format is to be discussed.

Note however that, even if all the item requests target a single shard, the index names can be different because each can refer a different alias that points to the same index. Also note that the `BulkShardRequest`s, unlike the `BulkItemRequest`s, might contain date math expressions for the index name:

2) 
```
curl -u elastic:password -X POST "localhost:9200/_bulk?pretty" -H 'Content-Type: application/json' -d'
{ "index" : { "_index" : "<index-{now/d}>", "_id" : "1" } }
{ "field1" : "value1" }
{ "index" : { "_index" : "<alias-{now/d+1d}>", "_id" : "1" } }
{ "field1" : "value1" }
{ "index" : { "_index" : "<index-{now/d+2d}>", "_id" : "1" } }
{ "field1" : "value1" }

{ "delete" : { "_index" : "<alias-{now/d}>", "_id" : "2" } }
{ "delete" : { "_index" : "<index-{now/d+1d}>", "_id" : "2" } }
{ "delete" : { "_index" : "<alias-{now/d+2d}>", "_id" : "2" } }

{ "create" : { "_index" : "<index-{now/d}>", "_id" : "2" } }
{ "field1" : "value2" }
{ "create" : { "_index" : "<alias-{now/d+1d}>", "_id" : "2" } }
{ "field1" : "value2" }
{ "create" : { "_index" : "<index-{now/d+2d}>", "_id" : "2" } }
{ "field1" : "value2" }

{ "update" : {"_id" : "1", "_index" : "<alias-{now/d}>"} }
{ "doc" : {"field2" : "value2"} }
{ "update" : {"_id" : "1", "_index" : "<index-{now/d+1d}>"} }
{ "doc" : {"field2" : "value2"} }
{ "update" : {"_id" : "1", "_index" : "<alias-{now/d+2d}>"} }
{ "doc" : {"field2" : "value2"} }
'
```
which is a similar bulk index operation that really targets 3 indices, but sometimes through an alias (there are 3 indices and 3 aliases with a 1-1 association), generates the following:
```
{... "event.action":"access_granted", ... , "action":"indices:data/write/bulk", "request.name":"BulkRequest"}

{... "event.action":"access_granted", ... , "action":"indices:data/write/bulk[s]", "request.name":"BulkShardRequest", "indices":["<index-{now/d+1d}>","<alias-{now/d+1d}>"]}
{... "event.action":"access_granted", ... , "action":"indices:data/write/index:op_type/index", "request.name":"BulkItemRequest", "indices":["alias-2020.12.18"]}
{... "event.action":"access_granted", ... , "action":"indices:data/write/delete", "request.name":"BulkItemRequest", "indices":["index-2020.12.18"]}
{... "event.action":"access_granted", ... , "action":"indices:data/write/index:op_type/create", "request.name":"BulkItemRequest", "indices":["alias-2020.12.18"]}
{... "event.action":"access_granted", ... , "action":"indices:data/write/update", "request.name":"BulkItemRequest", "indices":["index-2020.12.18"]}
{... "event.action":"access_granted", ... , "action":"indices:data/write/bulk[s][p]", "request.name":"BulkShardRequest", "indices":["<index-{now/d+1d}>","<alias-{now/d+1d}>"]}

{... "event.action":"access_granted", ... , "action":"indices:data/write/bulk[s]", "request.name":"BulkShardRequest", "indices":["<alias-{now/d+2d}>","<index-{now/d+2d}>"]}
{... "event.action":"access_granted", ... , "action":"indices:data/write/index:op_type/index", "request.name":"BulkItemRequest", "indices":["index-2020.12.19"]}
{... "event.action":"access_granted", ... , "action":"indices:data/write/delete", "request.name":"BulkItemRequest", "indices":["alias-2020.12.19"]}
{... "event.action":"access_granted", ... , "action":"indices:data/write/index:op_type/create", "request.name":"BulkItemRequest", "indices":["index-2020.12.19"]}
{... "event.action":"access_granted", ... , "action":"indices:data/write/update", "request.name":"BulkItemRequest", "indices":["alias-2020.12.19"]}
{... "event.action":"access_granted", ... , "action":"indices:data/write/bulk[s][p]", "request.name":"BulkShardRequest", "indices":["<alias-{now/d+2d}>","<index-{now/d+2d}>"]}

{... "event.action":"access_granted", ... , "action":"indices:data/write/bulk[s]", "request.name":"BulkShardRequest", "indices":["<index-{now/d}>","<alias-{now/d}>"]}
{... "event.action":"access_granted", ... , "action":"indices:data/write/index:op_type/index", "request.name":"BulkItemRequest", "indices":["index-2020.12.17"]}
{... "event.action":"access_granted", ... , "action":"indices:data/write/delete", "request.name":"BulkItemRequest", "indices":["alias-2020.12.17"]}
{... "event.action":"access_granted", ... , "action":"indices:data/write/index:op_type/create", "request.name":"BulkItemRequest", "indices":["index-2020.12.17"]}
{... "event.action":"access_granted", ... , "action":"indices:data/write/update", "request.name":"BulkItemRequest", "indices":["alias-2020.12.17"]}
{... "event.action":"access_granted", ... , "action":"indices:data/write/bulk[s][p]", "request.name":"BulkShardRequest", "indices":["<index-{now/d}>","<alias-{now/d}>"]}
```

Here's the interesting part of the code related to this https://github.com/elastic/elasticsearch/blob/b622adeb7adfe3dc0d731352e43075e94e464ff1/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java#L557

### PR DESCRIPTION
**Pull request by albertzaharovits**
This PR reduces the number of audit entries for bulk items in bulk shard requests.
Currently, each bulk item from every bulk shard request generates exactly one audit log entry.
This change reduces that to one audit log per action type (out of 4 possible) per bulk shard request.

For example:
```
curl -k -u "elastic:$ELASTIC_PASSWORD" -X POST "https://localhost:9200/_bulk?pretty" -H 'Content-Type: application/json' -d'
{ "index" : { "_index" : "<index-{now/d}>", "_id" : "1" } }
{ "field1" : "value1" }
{ "index" : { "_index" : "<alias-{now/d}>", "_id" : "2" } }
{ "field1" : "value2" }

{ "delete" : { "_index" : "<alias-{now/d}>", "_id" : "1" } }
{ "delete" : { "_index" : "<index-{now/d}>", "_id" : "2" } }

{ "create" : { "_index" : "<index-{now/d}>", "_id" : "1" } }
{ "field1" : "value1" }
{ "create" : { "_index" : "<alias-{now/d}>", "_id" : "2" } }
{ "field1" : "value2" }

{ "update" : {"_id" : "1", "_index" : "<alias-{now/d}>"} }
{ "doc" : {"field2" : "value1"} }
{ "update" : {"_id" : "2", "_index" : "<index-{now/d}>"} }
{ "doc" : {"field2" : "value2"} }
'
```
currently generates 8 audit log entries:
```
{... "action":"indices:data/write/index:op_type/index", "request.name":"BulkItemRequest", "indices":["index-2023.08.14"]}
{... "action":"indices:data/write/index:op_type/index", "request.name":"BulkItemRequest", "indices":["alias-2023.08.14"]}
{... "action":"indices:data/write/delete", "request.name":"BulkItemRequest", "indices":["alias-2023.08.14"]}
{... "action":"indices:data/write/delete", "request.name":"BulkItemRequest", "indices":["index-2023.08.14"]}
{... "action":"indices:data/write/index:op_type/create", "request.name":"BulkItemRequest", "indices":["index-2023.08.14"]}
{... "action":"indices:data/write/index:op_type/create", "request.name":"BulkItemRequest", "indices":["alias-2023.08.14"]}
{... "action":"indices:data/write/update", "request.name":"BulkItemRequest", "indices":["alias-2023.08.14"]}
{... "action":"indices:data/write/update", "request.name":"BulkItemRequest", "indices":["index-2023.08.14"]}

```
but following this PR it will change to 4 entries only:
```
{... "action":"indices:data/write/update", "request.name":"BulkItemRequest", "indices":["index-2023.08.16","alias-2023.08.16"]}
{... "action":"indices:data/write/index:op_type/create", "request.name":"BulkItemRequest", "indices":["index-2023.08.16","alias-2023.08.16"]}
{... "action":"indices:data/write/delete", "request.name":"BulkItemRequest", "indices":["index-2023.08.16","alias-2023.08.16"]}
{... "action":"indices:data/write/index:op_type/index", "request.name":"BulkItemRequest", "indices":["index-2023.08.16","alias-2023.08.16"]}
```

The reduction in log volume can be spectacular for bulk indexing requests with many items, especially if they're going into few shards and there are only few action types used (out of the 4 total).

Closes: https://github.com/elastic/elasticsearch/issues/66568

### CODE REVIEW

**Comment by albertzaharovits** [comment]
https://gradle-enterprise.elastic.co/s/rgixiouu456vk is https://github.com/elastic/elasticsearch/issues/98127

**Comment by albertzaharovits** [comment]
@elasticsearchmachine run https://github.com/elastic/elasticsearch/issues/98127

**Comment by albertzaharovits** [comment]
@elasticsearchmachine run elasticsearch-ci/part-3

**Comment by albertzaharovits** [comment]
@elasticmachine run elasticsearch-ci/part-1

**Comment by albertzaharovits** [comment]
Raised https://github.com/elastic/elasticsearch/issues/98539 for https://gradle-enterprise.elastic.co/s/6apv3wjzv26dw .

**Comment by albertzaharovits** [comment]
ping @n1v0lg

**Review by n1v0lg — COMMENTED** [review]
Code level changes LGTM!

One sanity check around behavior: before this PR, the audit log told us how many operations there were, albeit very inefficiently, by virtue of there being one `BulkItemRequest` entry per item. 

Now, that information is lost, e.g.,:

```
curl -k -u "elastic-admin:elastic-password" -X POST "http://localhost:9200/_bulk?refresh=true" -H 'Content-Type: application/json' -d'
{ "index" : { "_index" : "<index-2-{now/d}>", "_id" : "1" } }
{ "field1" : "value1" }
{ "index" : { "_index" : "<index-2-{now/d}>", "_id" : "2" } }
{ "field1" : "value2" }
```

Produces a single `BulkItemRequest` entry. I wonder if it makes sense to add a new field `item_count` or similar to indicate how many items there were.

Apologies if I missed something and this information is actually captured.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/AuditTrail.java [hunk 1]
```diff
@@ -104,7 +104,7 @@ void explicitIndexAccessEvent(
         AuditLevel eventType,
         Authentication authentication,
         String action,
-        String indices,
+        String[] indices,
```
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/AuditTrail.java:107]: It doesn't look like it. I've just searched for "explicitIndexAccess" in the `CustomAuthorizationEngine` class, and got no hits.

**Review by n1v0lg — APPROVED** [review]
LGTM -- as you point out we can always add the count in a follow-up, if we decide it's worthwhile.

**Pull request merged**